### ISSUE TITLE
Argument spread is not compiled correctly in case of excessive parens usage

### ISSUE THREAD

**Issue by andreypopp** [babel/babel#4634]
### Input Code

```js
let oneOf = (...nodes) => {
  if (nodes.length === 1) {
    return nodes[0];
  }
  return ((new OneOfNode(nodes)): any)
}
```

Results in

```js
var oneOf = function oneOf() {
  if (arguments.length === 1) {
    return arguments.length <= 0 ? undefined : arguments[0];
  }
  return new OneOfNode(nodes);
};
```

### Babel Configuration (.bablerc, package.json, cli command)

See [the snippet](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=false&presets=es2015%2Creact&code=let%20oneOf%20%3D%20(...nodes)%20%3D%3E%20%7B%0A%20%20if%20(nodes.length%20%3D%3D%3D%201)%20%7B%0A%20%20%20%20return%20nodes%5B0%5D%3B%0A%20%20%7D%0A%20%20return%20((new%20OneOfNode(nodes))%3A%20any)%0A%7D) on try babel.

```js
{
  "presets": ["es2015", "react"]
}
```

### Expected Behavior

Code does not fail at runtime.

### Current Behavior

Code fails at runtime because `nodes` are not defined.

### Your Environment

| software         | version
| ---------------- | -------
| Babel            |  6.16.0
| node             | 6.4.0
| npm              | 3.10.3
| Operating System | OS X El Capitan

**Comment by hzoo** [comment]
It looks like it is due to the wrap of `((new OneOfNode(nodes)): any)` in `(: any)`

**Comment by danez** [comment]
Should be fixed with #4636

**Comment by andreypopp** [comment]
Thank you!
